## Tuning Logistic Regression with GridSearchCV

In [1]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
import numpy as np
from sklearn.datasets import make_classification

In [6]:
X,y = make_classification(
    n_samples=1000,
    n_features=20,
    n_informative=10,
    n_classes=2,
    random_state=42
)

c_space = np.logspace(-5,8,15)
param_grid = {'C':c_space}

LR = LogisticRegression()
LR_cv = GridSearchCV(LR, param_grid, cv=5)
LR_cv.fit(X,y)

print('Tuned Logistic Regression Parameters: ',LR_cv.best_params_)
print('Best score is: ',LR_cv.best_score_)
print('Accuracy: ',LR_cv.best_score_*100)

Tuned Logistic Regression Parameters:  {'C': np.float64(0.006105402296585327)}
Best score is:  0.853
Accuracy:  85.3


This represents the highest accuracy achieved by the model using the hyperparameter combination,   ...C = 0.0061. The best score of 0.853 means the model achieved 85.3% accuracy on the validation data during the grid search process.

## 2. RandomizedSearchCV 

In [7]:
import numpy as np
from sklearn.datasets import make_classification

X,y = make_classification(
    n_samples = 1000,
    n_features=20,
    n_informative=10,
    n_classes=2,
    random_state=42
)

from scipy.stats import randint
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import RandomizedSearchCV

In [9]:
param_dist = {
    "max_depth":[3,None],
    "max_features": randint(1,9),
    "min_samples_leaf": randint(1,9),
    "criterion": ['gini', 'entropy']
}

tree = DecisionTreeClassifier()
tree_cv = RandomizedSearchCV(tree, param_dist, cv=5)
tree_cv.fit(X,y)

print("Tuned Decision Tree Parameters:", tree_cv.best_params_)
print('Best score is:', tree_cv.best_score_)


Tuned Decision Tree Parameters: {'criterion': 'gini', 'max_depth': None, 'max_features': 7, 'min_samples_leaf': 1}
Best score is: 0.8320000000000001


A score of 0.842 means the model performed with an accuracy of 84.2% on the validation set with following hyperparameters.

## 3. Bayesian Optimization

P(score(y)∣hyperparameters(x))

It treats hyperparameter tuning like a mathematical optimization problem and learns from past results to decide what to try next.

- Build a probabilistic model (surrogate function) that predicts performance based on hyperparameters.
- Update this model after each evaluation.
- Use the model to choose the next best set to try.
- Repeat until the optimal combination is found. The surrogate function models: